### ISSUE TITLE
Assertion error when trying to take jacobian of projected gradient solution using projection_polyhedron 

### ISSUE THREAD

**Issue by Basant1861** [google/jaxopt#338]
The output of compute_pg is a vector . If I just evaluate f_x below at the desired input it works but the jacobian doesnt seem to work. I am not sure what is happening here.

```
@partial(jit, static_argnums=(0,))
 def proj(self,p,C):
     return projection_polyhedron(p,C,check_feasible = False)

@partial(jit, static_argnums=(0,))
def compute_pg(self,p,cx,cy,x_obs_inp,y_obs_inp,vx_obs_inp, vy_obs_inp):
        
        p = jnp.reshape(p,(jnp.shape(p)[0],1))
        b_obs = jnp.hstack((x_obs_inp+self.p_obs_x_ub,-(x_obs_inp-self.p_obs_x_lb),
                                        y_obs_inp+self.p_obs_y_ub,-(y_obs_inp-self.p_obs_y_lb)))

        b_obs = jnp.reshape(b_obs,(jnp.shape(self.C_obs)[0],1))
        pg = ProjectedGradient(fun=self.compute_obstacle_penalty_temp,projection=self.proj,jit=True)
        pg_sol = pg.run(p,hyperparams_proj=(self.A_obs,self.a_obstacle,self.C_obs,b_obs),c_x=cx,c_y=cy,vx_obs_inp=vx_obs_inp,vy_obs_inp=vy_obs_inp).params
        return pg_sol 

f_x = lambda cx: self.compute_pg(p[0,:],cx,best_cy.T,x_obs_inp[0,:],y_obs_inp[0,:],vx_obs_inp[0,:], vy_obs_inp[0,:])
dp_dcx = jax.jacobian(f_x)(best_cx.T)

```

This is the error I get:

  ```
File "/home/ims/ros2_ws/install/mpc_python/lib/python3.8/site-packages/mpc_python/mpc_expert_bilevel.py", line 592, in compute_bilevel
    dp_dcx = jax.jacobian(f_x)(best_cx.T)
  File "/home/ims/.local/lib/python3.8/site-packages/jaxopt/_src/implicit_diff.py", line 236, in solver_fun_bwd
    vjps = root_vjp(optimality_fun=optimality_fun, sol=sol,
  File "/home/ims/.local/lib/python3.8/site-packages/jaxopt/_src/implicit_diff.py", line 69, in root_vjp
    u = solve(matvec, v)
  File "/home/ims/.local/lib/python3.8/site-packages/jaxopt/_src/linear_solve.py", line 193, in solve_normal_cg
    Ab = rmatvec(b)  # A.T b
  File "/home/ims/.local/lib/python3.8/site-packages/jaxopt/_src/linear_solve.py", line 145, in <lambda>
    return lambda y: transpose(y)[0]
AssertionError
```
What I observed was that if I set implicit_diff=False in the ProjectedGradient then it works but is super slow.Kindly advice.

**Comment by Algue-Rythme** [comment]
Hi Basant1861

When you differentiate a Jaxopt solver it will attempt - whenever possible - to differentiate with implicit differentiation. Implicit Differentiation is only possible if the argument you are trying to differentiate is part of the optimality conditions of your problem.

That's your issue; you are trying to differentiate with respect to `p` (i.e the initialization of ProjectedGradient) but it does not appear in the optimality conditions: for a convex problem `p` doesn't play any role. Hence mathematically the derivative should be zero anyway. Numerically, it's trickier.

Indeed, with implicit differentiation it does not work because Jaxopt cannot handle differentiating with respect to variables that are not part of optimality conditions. Without implicit differentiation, unrolling can return a value. This value should be zero in an ideal world, but with numerical errors I cannot guarantee it (I just tried and it is around `1e-9`). If you want to speed it up, you can try wrapping your whole `compute_pg` function in `jax.jit` and disable implicit diff.

I suggest you take a step back a think about the meaning of the derivative you want to compute. For example, on non-convex problems the initialization has an importance because different initialization will yield different (local) optima. But in this case the function that maps `p_0` to `p_t` (the optimum) is usually piecewise constant (each piece corresponding to a different basin of attraction).
